### PR TITLE
Svwzioep notion 509

### PR DESCRIPTION
**Pull request by beastoin**

### CODE REVIEW

File: plugins/example/client.py
Comment (beastoin) [inline plugins/example/client.py:1]: Remove this redundant, moved to oauth_client

File: plugins/example/main.py [hunk 1]
```diff
@@ -20,9 +23,9 @@
 
 @modal_app.function(
     image=(
-            Image.debian_slim()
-            # .apt_install('libgl1-mesa-glx', 'libglib2.0-0')
```
Comment (beastoin) [inline plugins/example/main.py]: revoke changes, formating

File: plugins/instructions/notion-crm-oauth/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#### --
+
+> If you have any feedback, please let us know at ngocthinhdp@gmail.com
+# Setting Up [Your Plugin Name]
```
Comment (beastoin) [inline plugins/instructions/notion-crm-oauth/README.md]: where is the name ?

File: plugins/instructions/notion-crm-oauth/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+   ![Step 1](assets/step_11.png)
+   ![Step 2](assets/step_12.png)
+
+2. Select Database you had copied from template above
+   ![Step 3](assets/step_13.png)
+   ![Step 4](assets/step_14.png)
+
+## Troubleshooting
+
+[Common issues and solutions]
```
Comment (beastoin) [inline plugins/instructions/notion-crm-oauth/README.md]: remove this line, there have no issue now

**Comment by josancamon19** [comment]
Hi @beastoin this shouldn't be a new plugin. Please check the current plugins/example/auth/ folder in the repo.

There's an existing notion plugin, this should be a completion of that plugin.

**Comment by beastoin** [comment]
> (This should be in another folder inside example/oauth) ~ but the community-plugins.json file should be updated to point to this new endpoint.

Sorry, got what your point of view (think so). So we don't need to create new community plugins (community-plugins.json), just need to point the current to the new one (oauth), right ?

---

> Hi @beastoin this shouldn't be a new plugin. Please check the current plugins/example/auth/ folder in the repo.
> 
> There's an existing notion plugin, this should be a completion of that plugin.

Hi @josancamon19,

Do you mean `plugins/example/basic_auth/` :? Actually, I used the pattern of the current Notion plugin in `plugins/example/basic_auth/`, then added the OAuth method and a prefix (/notion/) for the new OAuth Notion to ensure it does not break the current implementation.

So, should I update directly to basic_auth next?

**Comment by beastoin** [comment]
Hi Joan,

A new commit just came in. Basically, we now just need to maintain one Notion plugin in `plugins/example/oauth`. The APIs will look like this:

Screenshot 2024-08-08 at 16 46 36

Feel free to ask for changes if needed ~
Good day!

**Comment by beastoin** [comment]
Forgot to mention @josancamon19 , in case you need to discuss via chat, feel free to ping me on Discord at `@thinh`

> Hi Joan,
> 
> A new commit just came in. Basically, we now just need to maintain one Notion plugin in `plugins/example/oauth`. The APIs will look like this:
> Screenshot 2024-08-08 at 16 46 36
> 
> Feel free to ask for changes if needed ~ Good day!

**Comment by josancamon19** [comment]
Can you submit a video of this functionality?
Could be either from postman, browser or anywhere, thanks!

**Comment by josancamon19** [comment]
@beastoin

**Comment by beastoin** [comment]
Ok sure, please wait me for 15m @josancamon19

**Comment by beastoin** [comment]
Hi @josancamon19 new UAT just dropped, please take a look.

**UAT:**

1. Video demo (user view): 

https://github.com/user-attachments/assets/9277e8ab-9094-41ea-baa0-615f9b8a9195

2. System database:
Screenshot 2024-08-09 at 14 27 15

3. You can try live version at modal.com https://beastoin--plugins-examples-plugins-app.modal.run/setup-notion-crm?uid=thinhcd9-14b4-4656-9b1f-30823d7dab3a

Good day :)

File: plugins/example/oauth/memory_created.py [hunk 4]
```diff
@@ -63,15 +120,43 @@ def notion_crm(memory: Memory, uid: str):
...
 
     create_notion_row(notion_api_key, get_notion_database_id(uid), memory)
+
     return {}
 
 
 def create_notion_row(notion_api_key: str, database_id: str, memory: Memory):
-    # TODO: validate table exists and has correct fields
+    # Validate table exists and has correct fields
+    databaseOk = getNotion().getDatabase(database_id, notion_api_key)
```
Comment (josancamon19) [inline plugins/example/oauth/memory_created.py]: Please do snake case naming convention everywhere.

File: plugins/example/oauth/memory_created.py [hunk 5]
```diff
@@ -7,22 +7,78 @@
...
+    if "error" in databasesOk:
+        err = databasesOk["error"]
+        print(err)
+        return response_setup_notion_crm_page(request, uid, f"Something went wrong. Please try again! \n (code: 400003)")
+
+    # Pick top
+    databases = databasesOk["result"]
+    if len(databases) == 0 or databases[0].id == "":
+        return response_setup_notion_crm_page(request, uid, f"There is no database. Please try again!  \n (code: 400004)")
+    database_id = databases[0].id
```
Comment (josancamon19) [inline plugins/example/oauth/memory_created.py:72]: Can we include the db field validation here?

File: plugins/example/oauth_client.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
+import os
```
Comment (josancamon19) [inline plugins/example/oauth_client.py]: Let's move this to oauth folder

File: plugins/example/oauth/memory_created.py (hunk 4 above)
Comment (beastoin) [reply plugins/example/oauth/memory_created.py]: Yes, done. My fault 😵

File: plugins/example/oauth/memory_created.py [hunk 7]
```diff
@@ -7,22 +7,78 @@
...
+
+    # Save
+    print({'uid': uid, 'api_key': access_token, 'database_id': database_id})
+    store_notion_crm_api_key(uid, access_token)
+    store_notion_database_id(uid, database_id)
+    return templates.TemplateResponse("okpage.html", {"request": request, "uid": uid})
+
+
+@router.post('/creds/notion-crm', response_class=HTMLResponse, tags=['oauth'])
```
Comment (josancamon19) [inline plugins/example/oauth/memory_created.py]: do we still need this page?

Comment (josancamon19) [reply plugins/example/oauth/memory_created.py]: this endpoint*?

**Review by josancamon19 — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/7f858d14-4e4a-4368-931c-052977db8b02


Besides those changes, it seems to be working properly!

**Comment by josancamon19** [comment]
There's a template option https://josancamon19.notion.site/a391a37323a74dbaa798fc02e8d61006?v=25ae8d3a8ff4471ba5ecccda4bd89425&pvs=4

But when I oauth and select the "Provided template, doesn't seem to work

File: plugins/example/oauth/memory_created.py (hunk 5 above)
Comment (beastoin) [reply plugins/example/oauth/memory_created.py:72]: yes, of course

**Comment by beastoin** [comment]
> There's a template option https://josancamon19.notion.site/a391a37323a74dbaa798fc02e8d61006?v=25ae8d3a8ff4471ba5ecccda4bd89425&pvs=4
> 
> But when I oauth and select the "Provided template, doesn't seem to work

Seems like the code was oudated v/s template, let me update 😀

Screenshot 2024-08-10 at 05 33 04

**Comment by beastoin** [comment]
Resolved/all. We are good now, please take a look @josancamon19 

Screen Shot 2024-08-10 at 06 15 15

Screenshot 2024-08-10 at 06 13 03

**Review by josancamon19 — APPROVED** [review]

**Comment by kodjima33** [comment]
@beastoin Hey bro! We've sent bounty :) Thank you a lot for contributing! 

Can we chat asap? really realy curious to learn more about u! Just sent you a booking link and my phone number via email

**Pull request merged**